### PR TITLE
added statistics on single visit users - should they be ignored?

### PR DESCRIPTION
**Pull request by steremma**
Checking whether it makes sense to naively predict 0 revenue for users with a single visit (that was not a purchase)

### CODE REVIEW

File: google_analytics/notebooks/load.ipynb [hunk 1]
```diff
@@ -34,24 +34,189 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 65,
    "metadata": {},
    "outputs": [],
    "source": [
-    "train = load(\"../data/train.csv\", nrows=100000)\n",
-    "test = load(\"../data/test.csv\", nrows=10000)\n",
+    "x_train = train[(train[\"date\"] < '2016-12-01')]\n",
```
Comment (Christiannewisse) [inline google_analytics/notebooks/load.ipynb]: Our X_test data is from August 2017 until Oktober 15 2018.  Shouldn't our X_train be the same time period, but then the year before to make it a fair comparison? So, in this case, the period of X_train is August 2016 until Oktober 15, 2017 (date < 2017-10-15).

File: google_analytics/notebooks/load.ipynb [hunk 2]
```diff
@@ -34,24 +34,189 @@
...
+    "y_train = train[(train[\"date\"] >= '2016-12-01') & (train[\"date\"] >= '2017-02-01')]"
```
Comment (Christiannewisse) [inline google_analytics/notebooks/load.ipynb]: This relates to the comment above. If this comment is true, then y_train should be december 2017 and january 2018 (>=2017-12-01 and < '2018-02-01').

Comment (Christiannewisse) [reply google_analytics/notebooks/load.ipynb]: However, I just realize that when we want to check this for train_x (August 2016 till Oktober 15, 2017) and train_y (December 2017 and January 2018), we cannot do this yet. This is because, at this moment, we only have train data from August 2016 until August 2017. Is this the reason why you checked it for December 2016 and January 2017? If this is true, then I think this is indeed the best way to validate the assumption now. Maybe we can consider to re-evaluate it when we receive the new dataset?

File: google_analytics/notebooks/load.ipynb [hunk 3]
```diff
@@ -34,24 +34,189 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are 2290 people appearing both in the training and the test periods\n",
+      "We will ignore 2037 out of those 2290 people\n"
+     ]
+    }
+   ],
+   "source": [
+    "# We will only focus on people who exist in both datasets\n",
```
Comment (Christiannewisse) [inline google_analytics/notebooks/load.ipynb]: Am I correct that the question we want to answer is: How many people visited once during x_train and did not buy there (people in the list to_ignore) buy during y_train? The lower this number the better we validate our assumption to drop these people?
So I guess the reason why we focus on people existing in both datasets is that people in to_ignore that are not in y_train are definitely not buying?

File: google_analytics/notebooks/load.ipynb (hunk 2 above)
Comment (steremma) [reply google_analytics/notebooks/load.ipynb]: Exactly. You are correct, these will be the final timelines when the new dataset is released. But the current timelines is the best estimation I could think of using the current dataset.

File: google_analytics/notebooks/load.ipynb [hunk 4]
```diff
@@ -34,24 +34,189 @@
...
+     "output_type": "stream",
+     "text": [
+      "We made 196 mistakes (9.62% of cases) leading to loss = 5.4433800738203315\n"
+     ]
+    }
+   ],
+   "source": [
+    "import numpy as np\n",
+    "\n",
+    "predicted_target = [0] * len(common_ids)\n",
```
Comment (Christiannewisse) [inline google_analytics/notebooks/load.ipynb]: Were we set the predicted_target to zero for all common_ids. This means that also for people who are visiting and buying on a regular basis in y_train, we predict zero. I am correct with what I am saying here? Since I think this is not what we want?

File: google_analytics/notebooks/load.ipynb [hunk 5]
```diff
@@ -34,24 +34,189 @@
...
+    "\n",
+    "y_train_grouped = y_train[[\"fullVisitorId\", \"transactionRevenue\"]].groupby(\"fullVisitorId\", as_index=False).sum()\n",
+    "actual_target = y_train_grouped[\"transactionRevenue\"].apply(lambda revenue: np.log(revenue + 1))\n",
+    "    \n",
+    "loss = np.sqrt(np.mean((predicted_target - actual_target)**2))\n",
+    "num_mistakes = sum(actual_target > 0)\n",
+    "\n",
+    "print(\"We made {} mistakes ({:.2f}% of cases) leading to loss = {}\".format(num_mistakes, 100*num_mistakes/len(ignored) ,loss))"
```
Comment (Christiannewisse) [inline google_analytics/notebooks/load.ipynb]: Very nice that you translated the misclassification into the actual error on the leaderboard!

File: google_analytics/notebooks/load.ipynb (hunk 3 above)
Comment (steremma) [reply google_analytics/notebooks/load.ipynb]: Yes that is indeed the correct question. What I have answered here is slightly different however, because I am only taking into account the intersection of clients (people appearing at least once in both train and test). The denominator in my division should probably be the number of people in train regardless if they appear in test (those who don’t obviously have a 0 revenue). Therefore the actual numbers should be much lower than the ones I report here.

You can fix yourself or I can do it when I get back (Sunday/Monday)

File: google_analytics/notebooks/load.ipynb (hunk 4 above)
Comment (steremma) [reply google_analytics/notebooks/load.ipynb]: My intention was to predict 0 for those that should be ignored and compute the loss incurred by them alone. I’m not sure I did it correctly (hard to see the code on the phone :P)

File: google_analytics/notebooks/load.ipynb (hunk 2 above)
Comment (Christiannewisse) [reply google_analytics/notebooks/load.ipynb]: Great, I think you are totally right indeed!

File: google_analytics/notebooks/load.ipynb (hunk 3 above)
Comment (Christiannewisse) [reply google_analytics/notebooks/load.ipynb]: No problem! I will change it.

File: google_analytics/notebooks/load.ipynb (hunk 4 above)
Comment (Christiannewisse) [reply google_analytics/notebooks/load.ipynb]: I will check it to be sure.

**Comment by san89** [comment]
The modifications here proposed were addressed in a different pull request by @Christiannewisse .
